File: ch03/src/cpp/spawn_join.cpp
#include <atomic>
#include <format>
#include <iostream>
#include <stdexcept>
#include <thread>

namespace {
std::atomic_int32_t atomic_x { 0 };

void f()
{
    const auto x = atomic_x.load(std::memory_order_relaxed);

    if (x != 1 && x != 2) {
        throw std::runtime_error("--- assertion failed ---");
    }
}

}

int main()
{
    constexpr auto kExperiments = 1'000'000;
    for (auto i { 0 }; i < kExperiments; ++i) {
        if (i % 1000 == 0) {
            std::cout << std::format("Progress: {:7}/{} ({:4.1F})\n",
                i, kExperiments,
                static_cast<float>(i * 100) / kExperiments);
        }

        atomic_x.store(1, std::memory_order_relaxed);
        atomic_x.store(2, std::memory_order_relaxed);
        auto t = std::thread { f };
        t.join();
        atomic_x.store(3, std::memory_order_relaxed);
    }

    std::cout << "All assertions satisfied\n";
    return 0;
}

File: ch03/src/bin/spawn_join.rs
use std::{
    sync::atomic::{AtomicI32, Ordering},
    thread,
};

static X: AtomicI32 = AtomicI32::new(0);

fn f() {
    let x = X.load(Ordering::Relaxed);
    assert!(x == 1 || x == 2);
}

fn main() {
    const EXPERIMENTS: i32 = 1_000_000;
    for i in 0..EXPERIMENTS {
        if i % 1000 == 0 {
            println!(
                "Progress: {:7} / {} ({:4.1}%)",
                i,
                EXPERIMENTS,
                (i * 100) as f32 / EXPERIMENTS as f32
            );
        }

        X.store(1, Ordering::Relaxed);
        let t = thread::spawn(f);
        X.store(2, Ordering::Relaxed);
        t.join().unwrap();
        X.store(3, Ordering::Relaxed);

        let x = X.load(Ordering::Relaxed);
        assert_eq!(x, 3);
    }
    println!("All assertions satisfied");
}
